reassembler.go

```diff
@@ -98,7 +99,7 @@ func (r *Reassembler) PushMessage(msg *auparse.AuditMessage) {
 // timestamp and sequence number. If parsing fails then an error will be
 // returned. See PushMessage.
 func (r *Reassembler) Push(typ auparse.AuditMessageType, rawData []byte) error {
-	msg, err := auparse.Parse(typ, string(rawData))
+	msg, err := auparse.Parse(typ, internal.ByteSlice2String(rawData))
```

Comment:
This conversion is immediately undone in `auparse.Parse`. I'm wondering if it wouldn't be better to have that function work on `[]byte` rather than going through this unsafe rigmarole. That would obviously be a breaking change.

Reply:
Sure `parseAuditHeader()` in `auparse.Parse(..)` creates a slice of bytes from the given string first. But as `RawData` in `AuditMessage` expects a string, I do think it is not worth introducing a breaking change here.

Reply:
The breaking change I was thinking of would replace most of the uses of `string` here with `[]byte`, but yes, I think I agree in general. I can actually see good reason to change the signature of `parseAuditHeader` to `func parseAuditHeader(line string) (time.Time, uint32, int, error)`.
```diff
diff --git a/auparse/auparse.go b/auparse/auparse.go
index ff18f6a..5a65cc6 100644
--- a/auparse/auparse.go
+++ b/auparse/auparse.go
@@ -18,7 +18,6 @@
 package auparse
 
 import (
-       "bytes"
        "errors"
        "fmt"
        "regexp"
@@ -196,7 +195,7 @@ func ParseLogLine(line string) (*AuditMessage, error) {
 func Parse(typ AuditMessageType, message string) (*AuditMessage, error) {
        message = strings.TrimSpace(message)
 
-       timestamp, seq, end, err := parseAuditHeader([]byte(message))
+       timestamp, seq, end, err := parseAuditHeader(message)
        if err != nil {
                return nil, err
        }
@@ -213,41 +212,41 @@ func Parse(typ AuditMessageType, message string) (*AuditMessage, error) {
 
 // parseAuditHeader parses the timestamp and sequence number from the audit
 // message header that has the form of "audit(1490137971.011:50406):".
-func parseAuditHeader(line []byte) (time.Time, uint32, int, error) {
+func parseAuditHeader(line string) (time.Time, uint32, int, error) {
        // Find tokens.
-       start := bytes.IndexRune(line, '(')
+       start := strings.IndexRune(line, '(')
        if start == -1 {
                return time.Time{}, 0, 0, errInvalidAuditHeader
        }
-       dot := bytes.IndexRune(line[start:], '.')
+       dot := strings.IndexRune(line[start:], '.')
        if dot == -1 {
                return time.Time{}, 0, 0, errInvalidAuditHeader
        }
        dot += start
-       sep := bytes.IndexRune(line[dot:], ':')
+       sep := strings.IndexRune(line[dot:], ':')
        if sep == -1 {
                return time.Time{}, 0, 0, errInvalidAuditHeader
        }
        sep += dot
-       end := bytes.IndexRune(line[sep:], ')')
+       end := strings.IndexRune(line[sep:], ')')
        if end == -1 {
                return time.Time{}, 0, 0, errInvalidAuditHeader
        }
        end += sep
 
        // Parse timestamp.
-       sec, err := strconv.ParseInt(string(line[start+1:dot]), 10, 64)
+       sec, err := strconv.ParseInt(line[start+1:dot], 10, 64)
        if err != nil {
                return time.Time{}, 0, 0, errInvalidAuditHeader
        }
-       msec, err := strconv.ParseInt(string(line[dot+1:sep]), 10, 64)
+       msec, err := strconv.ParseInt(line[dot+1:sep], 10, 64)
        if err != nil {
                return time.Time{}, 0, 0, errInvalidAuditHeader
        }
        tm := time.Unix(sec, msec*int64(time.Millisecond)).UTC()
 
        // Parse sequence.
-       sequence, err := strconv.ParseUint(string(line[sep+1:end]), 10, 32)
+       sequence, err := strconv.ParseUint(line[sep+1:end], 10, 32)
        if err != nil {
                return time.Time{}, 0, 0, errInvalidAuditHeader
        }
diff --git a/auparse/auparse_test.go b/auparse/auparse_test.go
index 71f0d74..ed10481 100644
--- a/auparse/auparse_test.go
+++ b/auparse/auparse_test.go
@@ -95,7 +95,7 @@ func TestNormalizeAuditMessage(t *testing.T) {
 }
 
 func TestParseAuditHeader(t *testing.T) {
-       ts, seq, end, err := parseAuditHeader([]byte(syscallMsg))
+       ts, seq, end, err := parseAuditHeader(syscallMsg)
        if err != nil {
                t.Fatal(err)
        }
@@ -450,7 +450,7 @@ func readGoldenFile(name string) ([]*StoredAuditMessage, error) {
 }
 
 func BenchmarkParseAuditHeader(b *testing.B) {
-       msg := []byte(syscallMsg)
+       msg := syscallMsg
        for i := 0; i < b.N; i++ {
                _, _, _, err := parseAuditHeader(msg)
                if err != nil {
```

Reply:
I have applied this suggestion :white_check_mark: